src/cluster_legacy.c

```diff
@@ -7029,9 +7029,29 @@ int clusterCommandSpecial(client *c) {
         clusterDelNode(n);
         clusterDoBeforeSleep(CLUSTER_TODO_UPDATE_STATE | CLUSTER_TODO_SAVE_CONFIG);
         addReply(c, shared.ok);
-    } else if (!strcasecmp(c->argv[1]->ptr, "replicate") && c->argc == 3) {
-        /* CLUSTER REPLICATE <NODE ID> */
+    } else if (!strcasecmp(c->argv[1]->ptr, "replicate") && (c->argc == 3 || c->argc == 4)) {
+        /* CLUSTER REPLICATE (<NODE ID> | NO ONE)*/
         /* Lookup the specified node in our table. */
+        if (c->argc == 4) {
+            if (0 != strcasecmp(c->argv[2]->ptr, "NO") || 0 != strcasecmp(c->argv[3]->ptr, "ONE")) {
+                addReplySubcommandSyntaxError(c);
+                return 1;
+            }
+            if (nodeIsPrimary(myself)) {
+                addReply(c, shared.ok);
+                return 1;
+            }
+            serverLog(LL_NOTICE, "Stop replication and turning myself into empty primary.");
+            clusterSetNodeAsPrimary(myself);
+            clusterPromoteSelfToPrimary();
+            int empty_db_flags = server.repl_replica_lazy_flush ? EMPTYDB_ASYNC : EMPTYDB_NO_FLAGS;
+            emptyData(-1, empty_db_flags, NULL);
```

Comment:
To compare with the non-cluster command `REPLICAOF NO ONE`, that one does not empty the data.

In a cluster it's different, but why exactly? Do other cluster commands flush the data like this?

Reply:
In this case we should delete data because after detaching replica it is still in the cluster but has a role of empty primary. Thus, data should never be used. Any request will be redirected to other node in the cluster. In case of ```REPLICAOF NO ONE``` after disabling replication the node still can accept client requests and serve it with existing data.